### ISSUE TITLE
selinux prevents container access to USB serial devices inside (rootless) containers

### ISSUE THREAD

**Issue by mattdm** [containers/container-selinux#167]
I'm trying to run Home Assistant on Fedora IoT (and currently prototyping on Fedora Workstation). I need access to several USB serial devices, at `/dev/serial/by-id/usb-Silicon_Labs_HubZ_Smart_Home_Controller_71500102-if00-port0`. I did the thing with `--group-add keep-groups` and added the user to `dialout`, but... still errors. Oh look, an AVC. So, it works with `--security-opt label=disable`... but I don't want to do that. Instead I used `audit2allow` to do:

```
module container-usbtty 1.0;

require {
	type container_t;
	type usbtty_device_t;
	class chr_file { getattr ioctl lock open read write };
}

#============= container_t ==============
allow container_t usbtty_device_t:chr_file { getattr ioctl lock open read write };
```
Should there be a boolean? Or, more ideally, is there some way to be more fine-grained than this? Each device "belongs" to one container and nothing else should access it.

**Comment by rhatdan** [comment]
How are you adding the devices to the container?

--device ?

This works fine if working in rootfull mode.  In rootless mode we are just bind mounting the device into the container.

What is the risk of allowing containers to access usbtty devices.  Basically this means that any container would be allowed to attack any usb device, correct?

**Comment by rhatdan** [comment]
@wrabcak WDYT?  Is this something we should handle with udica?

**Comment by mattdm** [comment]
> How are you adding the devices to the container?
> 
> --device ?

Yep!

> This works fine if working in rootfull mode. In rootless mode we are just bind mounting the device into the container.

Yeah, I'm running rootless. I was working around this by using `--privileged` and `-v` with `:Z`, but that felt wrong. :) 

> What is the risk of allowing containers to access usbtty devices. Basically this means that any container would be allowed to attack any usb device, correct?

I think just tty (serial) USB devices. But yeah, I'd really prefer to bind them to the containers they're supposed to belong to. 

Right now, it's pretty theoretical because Home Assistant officially doesn't support anything but `--net=host` so there are worse problems.

**Comment by wrabcak** [comment]
Hi All, 

@mattdm, can you attach inspection json file of the container ? Feel free to do it with `--security-opt label=disable`. 

^^ This will tell us if udica would help here. 

Also, @vmojzis, this is probably material for new udica template, something like `--USB-access`, what do you think?

**Comment by rhatdan** [comment]
@wrabcak I would make it more global that that. Since --device-access, where it could look at the device, and allow access to the device type.

@mattdm you probably are not allowed to modify the device label from a rootless user, so you are stuck there.  With rootfull containers we actually create a new device based on the major/minor number of the device, private to the container and then label it correctly.  Rootless users are not allowed to create devices, so all we can do is bind mount them.
We could add a boolean like container_rw_devices which would allow containers to read/write all devices.  I would not want to make specific booleans to particular devices.  I would prefer udica to generate new types, which could be specific to a device.

**Comment by mattdm** [comment]
@rhatdan Attached!

[homeassistant-podman.json.gz](https://github.com/containers/container-selinux/files/7990586/homeassistant-podman.json.gz)
[zwave-podman.json.gz](https://github.com/containers/container-selinux/files/7990587/zwave-podman.json.gz)

> you probably are not allowed to modify the device label from a rootless user, so you are stuck there.

Yeah, that doesn't work. So I loaded the audit2allow-generated policy, but I always feel like I'm cheating if I find myself needing to do that. :)

**Comment by vmojzis** [comment]
> @rhatdan Attached!
> 
> [homeassistant-podman.json.gz](https://github.com/containers/container-selinux/files/7990586/homeassistant-podman.json.gz) [zwave-podman.json.gz](https://github.com/containers/container-selinux/files/7990587/zwave-podman.json.gz)

Hm, `"Devices": [],`, @rhatdan is this because the container is running rootless?

> 
> > you probably are not allowed to modify the device label from a rootless user, so you are stuck there.
> 
> Yeah, that doesn't work. So I loaded the audit2allow-generated policy, but I always feel like I'm cheating if I find myself needing to do that. :)

You could run each container with a custom label generated by udica and allow access to the devices individually (using `--append-rules <file_with_AVCs>` -- just like audit2allow, but for a single container). This way at least you are not loosening policy of all containers.
Custom labels for each device as Dan suggested should be doable. I'll look into it.

**Comment by rhatdan** [comment]
@vmojzis Yes in rootless containers we just bind mount the device into the container, which exposes the system device label to the container.

I have released container-selinux-2.176.0 which has a boolean to allow containers to use all devices, but having udica be able to generate types for specific containers using a specific device, would be nice.

**Comment by rhatdan** [comment]
Since there is a boolean now, I am going to close.
